### PR TITLE
Integrate CBGT for DCP sharding - issue #922

### PR DESCRIPTION
**Pull request by tleyden**

### CODE REVIEW

File: src/github.com/couchbase/sync_gateway/base/bucket.go [hunk 1]
```diff
@@ -121,10 +148,212 @@ func (bucket CouchbaseBucket) StartTapFeed(args sgbucket.TapArguments) (sgbucket
...
+	feedParams := bucket.spec.FeedParams
+
+	if bucket.spec.Auth != nil {
+		username, password, _ = bucket.spec.Auth.GetCredentials()
+	}
+
+	if username == "" {
+		username = bucket.Name
 	}
+
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket.go:207]: Can we reuse our bucket.getDcpAuthHandler() here to do the username="" handling?

File: src/github.com/couchbase/sync_gateway/base/bucket.go [hunk 2]
```diff
@@ -121,10 +148,212 @@ func (bucket CouchbaseBucket) StartTapFeed(args sgbucket.TapArguments) (sgbucket
...
+// Create an "index" in CBGT which will cause it to start streaming
+// DCP events to us for our shard of the full DCP stream.
+func (bucket CouchbaseBucket) createCBGTIndex() error {
+
+	authHandler := bucket.getDcpAuthHandler()
+	user, pwd, _ := authHandler.GetCredentials()
+	if user == "" && bucket.Name != "default" {
+		user = bucket.Name
+	}
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket.go]: dcpAuth.GetCredentials() is already doing some of this, but not the 'default' check.  If the name=default check is standard, we should move it into dcpAuth.GetCredentials().

File: src/github.com/couchbase/sync_gateway/base/sgw_pindex.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+			FailOverLog: failOver,
+		}
+		buf, err := json.Marshal(metadata)
+		if err == nil {
+			if s.meta == nil {
+				s.meta = make(map[uint16][]byte)
+			}
+			s.meta[vbucketId] = buf
+		}
+	}
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/sgw_pindex.go:116]: Are we still seeding with the high seqnos from the stats call?  Should be using the index bucket stable sequence - maybe that work is still pending?

File: src/github.com/couchbase/sync_gateway/base/bucket.go [hunk 4]
```diff
@@ -55,7 +78,9 @@ type couchbaseFeedImpl struct {
 	events <-chan sgbucket.TapEvent
 }
 
-var versionString string
+var (
+	versionString string
+)
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket.go]: Is this used?

**Comment by adamcfraser** [comment]
Looks good to me - just a few minor items/questions.

**Comment by tleyden** [comment]
@adamcfraser I made the changes, one of which required a fairly intensive refactor to move the SequenceClock into the base package.

I made the `SyncSequenceClock.Clock` field exported, since it was being accessed directly from `kv_change_index.go`:

```
// Returns the stable sequence clock for the index
func (k *kvChangeIndex) getStableClock() base.SequenceClock {
    return k.stableSequence.Clock
}
```

We could approach this differently, it felt a bit kludgy .. maybe keep it unexported and add an `ExportClock()` method?

Currently left as separate commits for easier viewing, will squash before merge.

**Comment by tleyden** [comment]
Dev build passed:

http://mobile.jenkins.couchbase.com/view/Sync_Gateway_Dev/job/sync-gateway-dev-build/261/

Ready for merge

**Pull request merged**